pytest-embedded-arduino/pytest_embedded_arduino/serial.py

```diff
@@ -40,33 +40,21 @@ def flash(self) -> None:
         """
         Flash the binary files to the board.
         """
-        flash_files = [
-            (offset, open(path, 'rb')) for (offset, path, encrypted) in self.app.flash_files if not encrypted
-        ]
+        flash_files = []
+        for offset, path, encrypted in self.app.flash_files:
+            if encrypted:
+                continue
+            flash_files.extend((str(offset), path))
 
-        default_kwargs = {
-            'addr_filename': flash_files,
-            'encrypt_files': None,
-            'no_stub': False,
-            'compress': True,
-            'verify': False,
-            'ignore_flash_encryption_efuse_setting': False,
-            'erase_all': False,
-            'encrypt': False,
-            'force': False,
-            'chip': self.app.target,
-        }
-
-        default_kwargs.update(self.app.flash_settings[self.app.target])
-        flash_args = EsptoolArgs(**default_kwargs)
+        flash_settings = []
+        for k, v in self.app.flash_settings[self.app.target].items():
+            flash_settings.append(f'--{k}')
+            flash_settings.append(v)
 
         try:
-            self.stub.change_baud(self.esptool_baud)
-            esptool.detect_flash_size(self.stub, flash_args)
-            esptool.write_flash(self.stub, flash_args)
-            self.stub.change_baud(self.baud)
+            esptool.main(
+                ['--chip', self.app.target, 'write_flash', *flash_files, *flash_settings, '--force'],
```

Comment:
Hi @radimkarnis. Here the esptool commands are with `--force` option. We've tested in the internal CI, and the pipeline won't pass without them. could you help take a final review?

Reply:
Hello guys,

`--force` is needed when:
1) The chip has flash encryption enabled
2) The chip has secure boot enabled
3) You are trying to flash an image for another chip revision (the chip revision is incompatible with the image)

ad 1) and 2) - I doubt you are using chips with active security features in the CI, so I guess the `--force` flag is needed for case 3). In that case, the better solution is to fix the image revision field in the image. Can you do that?

I am not a big fan of hard coding in the `--force` argument, it is there for a reason.

Reply:
@radimkarnis Thank you for the review. Now it's more clear to us.

For use case 1), we could improve it by

```python
if self.app.sdkconfig.get('CONFIG_SECURE_FLASH_ENC_ENABLED', False):
    args.append('--force')
```

For use case 2), we could improve it by

```python
if self.app.sdkconfig.get('CONFIG_SECURE_BOOT', False):
    args.append('--force')
```

as for use case 3), shouldn't we do it in `esptool` side? writing these logic inside `pytest-embedded` is quite weird. Or, is there any easier commands that we can call? if we could have `esptool flash --project-description-json <filepath>` that would be great.

I don't like the hard-coded `--force` either. For now I can suggest a quick workaround, let's add one `--esp-flash-force` CLI option insided `pytest-emebdded` for the replacement of use case 3). if this flag is specified, we'll add the `--force` while calling esptool flash.

Reply:
I have added the `--esp-flash-force` flag and a dependency on sdkconfig.

Reply:
ad 1) and 2) sound good to me! 

case 3): what exactly would you want to change in esptool? The correct behavior is to write correct tests (tests that use firmware images with the correct `min_rev` and `max_rev` fields - in this case, nothing needs to be changed in esptool and pytest-embedded).

But I am ok with the `--esp-flash-force` option, it adds flexibility.

LGTM!

Reply:
okay. I see. Thanks for the review Radim :)